src/models/StyledComponent.js

```diff
@@ -48,62 +51,111 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
+        }
+      }
+
       const executionContext = this.buildExecutionContext(theme, props)
       const styleSheet = this.context[CONTEXT_KEY] || StyleSheet.instance
       const className = componentStyle.generateAndInjectStyles(executionContext, styleSheet)
 
-      if (warnTooManyClasses !== undefined) warnTooManyClasses(className)
-
       return className
     }
 
     componentWillMount() {
       // If there is a theme in the context, subscribe to the event emitter. This
       // is necessary due to pure components blocking context updates, this circumvents
-      // that by updating when an event is emitted
-      if (this.context[CHANNEL]) {
-        const subscribe = this.context[CHANNEL]
-        this.unsubscribe = subscribe(nextTheme => {
-          // This will be called once immediately
-
-          // Props should take precedence over ThemeProvider, which should take precedence over
-          // defaultProps, but React automatically puts defaultProps on props.
-          const { defaultProps } = this.constructor
-          const isDefaultTheme = defaultProps && this.props.theme === defaultProps.theme
-          const theme = this.props.theme && !isDefaultTheme ? this.props.theme : nextTheme
-          const generatedClassName = this.generateAndInjectStyles(theme, this.props)
-          this.setState({ theme, generatedClassName })
+      // that by updating when an event is emitted.
+      // But remember, avoid allocation listeners and building extra work for ourselves if
+      // its static component
+      const { componentStyle } = this.constructor
+      if (componentStyle.isStaticRules) {
+        return
+      }
+
+      const styledContext = this.context[CHANNEL]
+      if (styledContext) {
+        const { subscribe } = styledContext
+        this.unsubscribeId = subscribe(nextTheme => {
+          if (nextTheme !== this.state.theme) {
+            // This will be called once immediately
+
+            // Props should take precedence over ThemeProvider, which should take precedence over
+            // defaultProps, but React automatically puts defaultProps on props.
+            const { defaultProps } = this.constructor
+            const isDefaultTheme = defaultProps && this.props.theme === defaultProps.theme
+            const theme = this.props.theme && !isDefaultTheme ? this.props.theme : nextTheme
+            const generatedClassName = this.generateAndInjectStyles(theme, this.props)
+            this.setState({ theme, generatedClassName })
+          }
         })
-      } else {
-        const theme = this.props.theme || {}
-        const generatedClassName = this.generateAndInjectStyles(
-          theme,
-          this.props,
-        )
-        this.setState({ theme, generatedClassName })
       }
     }
 
     componentWillReceiveProps(nextProps: { theme?: Theme, [key: string]: any }) {
+      const { componentStyle } = this.constructor
+      if (componentStyle.isStaticRules) {
+        return
```

Comment:
Ah, I see now why you added the branch for static styles here. Maybe we can remove it's extra logic from ComponentStyle and instead move some here. The separation is a bit blurry now. Maybe leave it as is now, since in v3 there'll be an extra case for static styles only anyway :+1: